Make raftstore testable with arbitrary combinations of storage engines

brson opened the pull request:
### What problem does this PR solve?

Issue Number: close https://github.com/tikv/tikv/issues/8452

Problem Summary:

Make it possible to run the test suite with alternate, non-rocksdb storage engines.

### What is changed and how it works?

This patch adds a crate, `engine_test`, that does two things:

- Adds compile-time features to select which storage engine type will be used for the kv store and which will be used for the raft store, both independent from each other
- Defines common constructors for all storage engines

Eventually the constructor code will be moved elsewhere, probably into `engine_traits` and individual engines. But for now they only fulfill the use cases needed by the raftstore crate, duplicate some `engine_traits` code, and I'd like to keep them in `engine_test` while they evolve, to indicate they are only useful for test use cases so far.

This patch then modifies the raftstore unit tests to use `engine_test`.

After this patch there are only a few minor places raftstore still uses `engine_rocks` directly:

- `PerfContext`-related code, which only affects global state and does nothing when other engines are selected
- Some config code
- A test case that uses the concrete `RocksCompactedEvent` type, that I haven't decided what to do about yet.

Other than that one test case, raftstore is completely testable with alternate storage engines. Of course, as of now the only alternate engine in tree is `engine_panic` which just panics, but it can be tested with:

```
cargo test -p raftstore --no-default-features --features=protobuf-codec,test-engines-panic
```

More details about how the `engine_test` crate should be used in the crate docs for `engine_test`.

This will eventually require more work to accommodate the special raft engine, but it should essentially be compatible with this design. cc @hicqu 

Next steps:

- Completely remove the engine_rocks dependency from raftstore, so that nobody can re-introduce unabstracted dependencies. The main obstacle here is figuring out what to do with PerfContext.
- Continue abstracting the tikv crate. It's mostly straightforward from here.
- I have begun working on `engine_sled`, which should be a pretty simple binding, to prove that it's possible to test other engines successfully.
- Get @5kbpers's `engine_skiplist` merged and make it testable.

cc @nrc you might look at the design of `engine_test` and see if you think it's maintainable, as it will complicate the testing story eventually with it's compile-time configuration.

### Related changes


### Check List <!--REMOVE the items that are not applicable-->


### Release note <!-- bugfixes or new feature need a release note -->

- No release note.

brson commented:
cc @yiwu-arbug i think you have been doing work on encrypted storage. This patch begins making storage engine constructors work across multiple storage engines, and adds a case for the most simple encrypted storage use case, adding a `DBOptions` method called `with_default_ctr_encrypted_env`.

I don't understand the `Env` type yet, or the encryption features, so you might peek at this and see if the direction I'm going is compatible with your work.

brson commented:
In general, the constructors for each engine are going to need to hide a lot of the details of how to construct the engines in a proper configuration for tikv.

So e.g. the constructors here just assume that you always want the `MvccPropertyCollectorFactory` and `RangePropertyCollectorFactory` installed, since those are pretty much necessary for the engine to work properly within tikv, and representing those types in an abstract way is probably not reasonable.

brson commented:
This is rebased and the makefile integration should be fixed.

brson commented:
ping @hicqu @Little-Wallace @5kbpers

5kbpers commented on components/engine_test/src/lib.rs:
```diff
@@ -0,0 +1,451 @@
+    pub fn new_engine(
+        path: &str,
+        db_opt: Option<DBOptions>,
+        cfs: &[&str],
```
Raft engine has only one cf so we shall accept only one cf option here.

brson replied:
I have made this change.

5kbpers reviewed (approved):
Just one comment, rest LGTM, thanks!

nrc commented on components/engine_test/src/lib.rs:
```diff
@@ -0,0 +1,451 @@
+
+/// Types and constructors for the "raft" engine
+pub mod raft {
+    use crate::ctor::{CFOptions, DBOptions, EngineConstructorExt};
```
style: should be CfOptions and DbOptions

brson replied:
This is also just carrying over existing names from the rocks apis and again I'd prefer to do API cleanup as a dedicated task elsewhere.

nrc commented on components/engine_test/src/lib.rs:
```diff
@@ -0,0 +1,451 @@
+        RaftTestEngine::new_engine(path, db_opt, cfs, opts)
+    }
+
+    pub fn new_engine_opt(
```
what does `_opt` mean here? Both functions return a Result (not Option) and both take both db and cf options

brson replied:
I think it does refer to the fact that the function takes options, but the messy API is legacy cruft. These are just wrappers, with the same names, around existing APIs. There really shouldn't be two constructor functions here that do almost the exact same thing, but for now I've been mostly preserving existing APIs to ease the transition. Eventually these should be cleaned up, but I'd rather not do that now.

nrc commented on components/engine_test/src/lib.rs:
```diff
@@ -0,0 +1,451 @@
+    }
+
+    #[derive(Clone)]
+    pub enum CryptoOpts {
```
I think CryptoOptions would be a better name - its only a few chars and means it follows db options and cf options.

brson replied:
Will do.

nrc commented on components/engine_test/src/lib.rs:
```diff
@@ -0,0 +1,451 @@
+    }
+
+    pub struct CFOptions<'a> {
+        pub cf: &'a str,
```
perhaps `name` instead of `cf`?

brson replied:
This is again mirroring pre-existing APIs and I'd like to leave as-is for now.

nrc reviewed (approved):
lgtm, just some style nits in comments

brson commented:
/merge

brson commented:
I've addressed all comments, but as indicated I have mostly declined making stylistic changes that alter the existing rocks API names, to maintain consistency with the existing APIs while they are still used in the rest of the codebase. The API should be reviewed for stylistic consistency but I prefer to do that sometime in the future.

brson commented:
/merge
